File: PKHImageCache/PKHImageCache.m
//
//  PKHImageCache.m
//  PKHImageCache
//
//  Created by Patrick Hanlon on 5/4/15.
//  Copyright (c) 2015 Patrick Hanlon. All rights reserved.
//

#import "PKHImageCache.h"

#import <CommonCrypto/CommonDigest.h>
#import "PKHImageDownloadOperation.h"

static const NSInteger kMaxCacheAgeForUnusedImages = 60 * 60 * 24 * 3; // 3 days
static const NSInteger kMaxCacheAgeForAllImages = 60 * 60 * 24 * 14; // 2 weeks

@interface PKHImageCache ()

@property (nonatomic, strong) NSCache *memoryCache;
@property (nonatomic) dispatch_queue_t pkhImageCacheQueue;
@property (nonatomic) NSInteger maxCacheAge;

@end

@implementation PKHImageCache
{
    NSFileManager *_fileManager;
}

+ (PKHImageCache *)sharedImageCache
{
    static dispatch_once_t once;
    static id instance;
    dispatch_once(&once, ^{
        instance = [self new];
    });
    return instance;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        _memoryCache = [[NSCache alloc] init];
        _memoryCache.name = @"default";
        
        _maxCacheAge = kMaxCacheAgeForAllImages;
        
        _pkhImageCacheQueue = dispatch_queue_create("io.pkh.PKHImageCache", DISPATCH_QUEUE_SERIAL);
        
        dispatch_sync(_pkhImageCacheQueue, ^{
            _fileManager = [NSFileManager new];
        });
        
        // Subscribe to app events
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(clearMemory)
                                                     name:UIApplicationDidReceiveMemoryWarningNotification
                                                   object:nil];
        
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(cleanDisk)
                                                     name:UIApplicationWillTerminateNotification
                                                   object:nil];
        
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(backgroundCleanDisk)
                                                     name:UIApplicationDidEnterBackgroundNotification
                                                   object:nil];
        
    }
    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark -

- (void)addImageOperationForImageView:(UIImageView *)imageView usingURL:(NSURL *)imageURL andPlaceholderImage:(UIImage *)placeholder
{
    // First, set the image view's placeholder image
    dispatch_async(dispatch_get_main_queue(), ^{
        imageView.image = placeholder;
    });
    
    if (!imageURL) {    // if imageURL is nil, skip the rest of this
        return;
    }
    
    __weak typeof(self)weakSelf = self;
    
    dispatch_async(self.pkhImageCacheQueue, ^{
        
        // Look for image in local cache
        UIImage *cachedImage = [weakSelf searchLocalImageCacheForImageURL:imageURL];
    
        if (cachedImage) {
            dispatch_async(dispatch_get_main_queue(), ^{
                imageView.image = cachedImage;
            });
        } else {
            // if image NOT present in local cache, enqueue download operation
            PKHImageDownloadOperation *operation = [PKHImageDownloadOperation new];
            operation.imageURLString = [imageURL absoluteString];
            operation.imageView = imageView;
            [operation start];
        }
    });
}

- (void)insertImageInLocalCache:(UIImage *)image withImageURLString:(NSString *)imageURLString
{
    NSString *hashedFilename = [self hashedImageURLString:imageURLString];
    
    NSString *cacheDirectoryPath = [self cacheDirectoryPath];
    
    dispatch_async(self.pkhImageCacheQueue, ^{
        
        if ([_fileManager fileExistsAtPath:cacheDirectoryPath isDirectory:NULL] == NO) {
            // create cache directory
            [_fileManager createDirectoryAtPath:cacheDirectoryPath withIntermediateDirectories:NO attributes:nil error:NULL];
        }
        
        NSString *diskImagePath = [cacheDirectoryPath stringByAppendingPathComponent:hashedFilename];
        
        [_fileManager createFileAtPath:diskImagePath contents:UIImagePNGRepresentation(image) attributes:nil];
        
        [self.memoryCache setObject:image forKey:hashedFilename];
        
    });
}

- (void)clearAndEmptyCache
{
    dispatch_async(self.pkhImageCacheQueue, ^{
        
        // clear the in-memory cache
        [self.memoryCache removeAllObjects];
        
        // clear the on-disk cache
        NSError *error = nil;
        [_fileManager removeItemAtPath:[self cacheDirectoryPath] error:&error];
        if (error != nil) {
            NSLog(@"Error removing disk cache: %@", error);
        }
        
        // Re-create the on-disk cache folder
        [_fileManager createDirectoryAtPath:[self cacheDirectoryPath] withIntermediateDirectories:NO attributes:nil error:NULL];
        
    });
}

#pragma mark - Private

- (UIImage *)searchLocalImageCacheForImageURL:(NSURL *)imageURL
{
    NSString *imageURLHash = [self hashedImageURLString:[imageURL absoluteString]];
    
    // First, search the in memory cache
    UIImage *cachedImage = [self.memoryCache objectForKey:imageURLHash];
    if (cachedImage) {
        NSLog(@"Have image in in-memory cache!");
        return cachedImage;
    }
    
    // Second, search the cache on disk
    NSData *data = [NSData dataWithContentsOfFile:[self diskImagePathWithHash:imageURLHash]];
    if (data) {
        NSLog(@"Have image in on-disk cache!");
        UIImage *image = [UIImage imageWithData:data];
        
        [self.memoryCache setObject:image forKey:imageURLHash];
        
        return image;
    }
    
    NSLog(@"Can't find image locally....");
    return nil;
}

#pragma mark - Helpers

- (NSString *)diskImagePathWithHash:(NSString *)hash
{
    return [[self cacheDirectoryPath] stringByAppendingPathComponent:hash];
}

- (NSString *)hashedImageURLString:(NSString *)imageURLString
{
    const char *charArrayStr = [imageURLString UTF8String];
    if (charArrayStr == NULL) {
        charArrayStr = "";
    }
    unsigned char r[CC_MD5_DIGEST_LENGTH];
    CC_MD5(charArrayStr, (CC_LONG)strlen(charArrayStr), r);
    NSString *hashedFilename = [NSString stringWithFormat:@"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
                                r[0], r[1], r[2], r[3], r[4], r[5], r[6], r[7], r[8], r[9], r[10], r[11], r[12], r[13], r[14], r[15]];
        
    return hashedFilename;
}

- (NSString *)cacheDirectoryPath
{
    NSString *cacheName = @"io.pkh.PKHImageCache";
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    return [paths[0] stringByAppendingPathComponent:cacheName];
}

- (NSUInteger)cacheSize
{
    __block NSUInteger size = 0;
    dispatch_sync(self.pkhImageCacheQueue, ^{
        NSDirectoryEnumerator *fileEnumerator = [_fileManager enumeratorAtPath:[self cacheDirectoryPath]];
        for (NSString *fileName in fileEnumerator) {
            NSString *filePath = [[self cacheDirectoryPath] stringByAppendingPathComponent:fileName];
            NSDictionary *attrs = [[NSFileManager defaultManager] attributesOfItemAtPath:filePath error:nil];
            size += [attrs fileSize];
        }
    });
    return size;
}

#pragma mark - Clean Up

- (void)clearMemory
{
    [self.memoryCache removeAllObjects];
}

- (void)cleanDisk
{
    [self cleanDiskWithCompletion:nil];
}

- (void)cleanDiskWithCompletion:(void(^)(void))completionBlock
{
    // clean up the on-disk cache by deleting image files
    // that haven't been accessed in 3 days (aggressive option)
    
    __weak typeof(self)weakSelf = self;
    
    dispatch_async(self.pkhImageCacheQueue, ^{
        
        NSDate *expirationDate = [NSDate dateWithTimeIntervalSinceNow:-self.maxCacheAge];
        NSLog(@"expirationDate: %@",expirationDate);
        
        NSURL *cacheURL = [NSURL fileURLWithPath:[weakSelf cacheDirectoryPath] isDirectory:YES];
        NSArray *resourceKeys = @[NSURLIsDirectoryKey, NSURLContentAccessDateKey, NSURLCreationDateKey];
        
        NSDirectoryEnumerator *fileEnumerator = [_fileManager enumeratorAtURL:cacheURL
                                                   includingPropertiesForKeys:resourceKeys
                                                                      options:NSDirectoryEnumerationSkipsHiddenFiles
                                                                 errorHandler:NULL];
        
        NSMutableSet *fileURLsToDelete = [NSMutableSet new];
        
        for (NSURL *fileURL in fileEnumerator) {
            
            NSDictionary *resourceValues = [fileURL resourceValuesForKeys:resourceKeys error:NULL];
            
            if ([resourceValues[NSURLIsDirectoryKey] boolValue]) {
                continue;
            }
            
            // if this image is older than 2 weeks, clear it from cache regardless
            NSDate *creationDate = resourceValues[NSURLCreationDateKey];
            if ([[creationDate laterDate:expirationDate] isEqualToDate:expirationDate]) {
                [fileURLsToDelete addObject:fileURL];
            }
            
            // if this image hasn't been accessed in the last 3 days, remove it from the cache
            NSDate *lastAccessDate = resourceValues[NSURLContentAccessDateKey];
            NSDate *recentlyUsedExpirationDate = [NSDate dateWithTimeIntervalSinceNow:-kMaxCacheAgeForUnusedImages];
            
            if ([[lastAccessDate laterDate:recentlyUsedExpirationDate] isEqualToDate:recentlyUsedExpirationDate]) {
                if ([fileURLsToDelete containsObject:fileURL] == NO) {
                    [fileURLsToDelete addObject:fileURL];
                }
            }
            
        }
        
        NSLog(@"Deleting %lu files",(long)[fileURLsToDelete count]);
        NSLog(@"%@",fileURLsToDelete);
        
        for (NSURL *fileURL in fileURLsToDelete) {
            [_fileManager removeItemAtURL:fileURL error:nil];
        }
        
        if (completionBlock) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completionBlock();
            });
        }
    });
}

- (void)backgroundCleanDisk
{
    UIApplication *app = [UIApplication sharedApplication];
    __block UIBackgroundTaskIdentifier bgTask = [app beginBackgroundTaskWithExpirationHandler:^{
        [app endBackgroundTask:bgTask];
        bgTask = UIBackgroundTaskInvalid;
    }];
    
    [self cleanDiskWithCompletion:^{
        [app endBackgroundTask:bgTask];
        bgTask = UIBackgroundTaskInvalid;
    }];
}

@end

File: PKHImageCache/PKHImageCache.swift
//
//  PKHImageCache.swift
//  PKHIC-Example
//
//  Created by Patrick Hanlon on 6/23/15.
//  Copyright (c) 2015 pkh. All rights reserved.
//

import Foundation
import UIKit

typealias PKHImageCacheCompletionBlock = (cachedImage: UIImage?, cachedImageURL: NSURL?) -> Void

extension UIImageView {
    
    @objc func pkhic_setImageWith(imageURL: NSURL?, placeholder: UIImage?) {
        
        self.image = placeholder!
        
        if (imageURL == nil) {
            return
        }
        
        weak var weakSelf = self
        
        PKHImageCache.sharedImageCache.addImageOperation(imageURL, completionBlock: { (cachedImage, cachedImageURL) -> Void in
            
            if (cachedImage != nil && imageURL?.absoluteString == cachedImageURL?.absoluteString) {
                weakSelf?.image = cachedImage
            }
            
        })
        
    }
    
}

class PKHImageCache : NSObject {
    
    static let sharedImageCache = PKHImageCache()
    
    private let kMaxCacheAgeForUnusedImages: Int = 60 * 60 * 24 * 3; // 3 days
    private let kMaxCacheAgeForAllImages: Int = 60 * 60 * 24 * 14; // 2 weeks
    
    private var memoryCache: NSCache
    private var pkhImageCacheQueue: dispatch_queue_t
    private var maxCacheAge: Int
    
    private let fileManager: NSFileManager = NSFileManager()
    
    // MARK:
    
    override init() {
        
        self.memoryCache = NSCache()
        self.memoryCache.name = "default"
        
        self.maxCacheAge = kMaxCacheAgeForAllImages;
        
        self.pkhImageCacheQueue = dispatch_queue_create("io.pkh.PKHImageCache", DISPATCH_QUEUE_SERIAL)
        /*
        dispatch_sync(queue: self.pkhImageCacheQueue) { () -> Void in
            self.fileManager = NSFileManager()
        }
        */
        super.init()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "clearMemory", name: UIApplicationDidReceiveMemoryWarningNotification, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "cleanDisk", name: UIApplicationWillTerminateNotification, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "backgroundCleanDisk", name: UIApplicationDidEnterBackgroundNotification, object: nil)
        
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // MARK:
    
    func addImageOperation(imageURL: NSURL?, completionBlock: PKHImageCacheCompletionBlock?) {
        
        if (imageURL == nil) {    // // if imageURL is nil, skip the rest of this
            if (completionBlock != nil) {
                completionBlock!(cachedImage: nil, cachedImageURL: nil)
            }
            return
        }
        
        dispatch_async(self.pkhImageCacheQueue, { () -> Void in
            
            // look for image in local cache
            let cachedImage: UIImage? = self.searchLocalImageCacheForImageURL(imageURL!)
            
            if (cachedImage != nil) {
                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                    completionBlock!(cachedImage: cachedImage!, cachedImageURL: imageURL!)
                })
            } else {
                // image NOT present in local cache, enqueue download operation
                let operation: PKHImageDownloadOperation = PKHImageDownloadOperation()
                operation.start(imageURL!.absoluteString!, completionBlock: { (image, imageURL) -> Void in
                    self.insertImageInLocalCache(image, urlString: imageURL!.absoluteString!)
                    
                    if let completion = completionBlock {
                        dispatch_async(dispatch_get_main_queue(), { () -> Void in
                            completion(cachedImage: image, cachedImageURL: imageURL!)
                        })
                    }
                    
                })
                
            }
            
        })
    }
    
    func insertImageInLocalCache(image: UIImage, urlString: String) {
        
        let hashedFileName = self.hashedImageURLString(urlString)
        let cacheDirectoryPath = self.cacheDirectoryPath()
        
        dispatch_async(self.pkhImageCacheQueue, { () -> Void in
            
            if (self.fileManager.fileExistsAtPath(cacheDirectoryPath) == false) {
                // create cache directory
                self.fileManager.createDirectoryAtPath(cacheDirectoryPath, withIntermediateDirectories: false, attributes: nil, error: nil)
            }
            
            let diskImagePath = cacheDirectoryPath.stringByAppendingPathComponent(hashedFileName)
            self.fileManager.createFileAtPath(diskImagePath, contents: UIImagePNGRepresentation(image), attributes: nil)
            self.memoryCache.setObject(image, forKey: hashedFileName)
        })
    
    }
    
    @objc func clearAndEmptyCache() {
        dispatch_async(self.pkhImageCacheQueue, { () -> Void in
            
            // clear the in-memory cache
            self.memoryCache.removeAllObjects()
            
            // clear the on-disk cache
            var error: NSError?
            self.fileManager.removeItemAtPath(self.cacheDirectoryPath(), error: &error)
            if (error != nil) {
                println("Error removing disk cache: \(error)")
            }
            
            // Re-create the on-disk cache folder
            self.fileManager.createDirectoryAtPath(self.cacheDirectoryPath(), withIntermediateDirectories: false, attributes: nil, error: nil)

        })
    }
    
    @objc func cacheSize() -> Int {
        var size: Int = 0;
        dispatch_sync(self.pkhImageCacheQueue, { () -> Void in
            let fileEnumerator: NSDirectoryEnumerator = self.fileManager.enumeratorAtPath(self.cacheDirectoryPath())!
            for fileName in fileEnumerator {
                let filePath: String = self.cacheDirectoryPath().stringByAppendingPathComponent(fileName as! String)
                let attrs: NSDictionary = self.fileManager.attributesOfItemAtPath(filePath, error: nil)!
                size += Int(attrs.fileSize())
            }
        });
        return size
    }
    
    // MARK: Private
    
    private func searchLocalImageCacheForImageURL(imageURL: NSURL) -> UIImage? {
        
        let imageURLHash = self.hashedImageURLString(imageURL.absoluteString!)
        
        // First, search the in-memory cache
        var cachedImage: UIImage? = self.memoryCache.objectForKey(imageURLHash) as? UIImage
        if (cachedImage != nil) {
            println("Have image in in-memory cache")
            return cachedImage
        }
        
        // Second, search the cache on disk
        var imageData: NSData? = NSData(contentsOfFile: self.diskImagePathWithHash(imageURLHash))
        if (imageData != nil) {
            println("Have image in on-disk cache!")
            let image = UIImage(data: imageData!)
            self.memoryCache.setObject(image!, forKey: imageURLHash)
            return image
        }
        
        println("Can't find image locally...")
        return nil
    }
    
    // MARK: Helpers
    
    private func diskImagePathWithHash(hash: String) -> String {
        return self.cacheDirectoryPath().stringByAppendingPathComponent(hash)
    }
    
    private func hashedImageURLString(imageURLString: String) -> String {
        
        let str = imageURLString.cStringUsingEncoding(NSUTF8StringEncoding)
        let strLen = CC_LONG(imageURLString.lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
        let digestLen = Int(CC_MD5_DIGEST_LENGTH)
        let result = UnsafeMutablePointer<CUnsignedChar>.alloc(digestLen)
        
        CC_MD5(str!, strLen, result)
        
        var hash = NSMutableString()
        for i in 0..<digestLen {
            hash.appendFormat("%02x", result[i])
        }
        
        result.dealloc(digestLen)
        
        return hash as String
        
    }
    
    private func cacheDirectoryPath() -> String {
        let cacheName: String = "io.pkh.PKHImageCache"
        let paths: NSArray = NSSearchPathForDirectoriesInDomains(.CachesDirectory, .UserDomainMask, true) as! [String]
        return paths[0].stringByAppendingPathComponent(cacheName)
    }
    
    // MARK: Clean Up
    
    private func clearMemory() {
        println("clearMemory")
        self.memoryCache.removeAllObjects()
    }
    
    private func cleanDisk() {
        println("cleanDisk")
        self.cleanDiskWithCompletion({ _ in })
    }
    
    private func cleanDiskWithCompletion(completion: ()->() ) {
        
        println("cleanDiskWithCompletion")
        
        // clean up the on-disk cache by deleting image files
        // that haven't been accessed in 3 days (aggressive option)
        
        dispatch_async(self.pkhImageCacheQueue, { () -> Void in
            
            let expirationDate: NSDate = NSDate().dateByAddingTimeInterval(NSTimeInterval(-self.maxCacheAge))
            println("expirationDate: \(expirationDate)")
            
            let cacheURL: NSURL = NSURL(fileURLWithPath: self.cacheDirectoryPath(), isDirectory: false)!
            let resourceKeys: [String] = [NSURLIsDirectoryKey, NSURLContentAccessDateKey, NSURLCreationDateKey]
            
            let fileEnumerator: NSDirectoryEnumerator = self.fileManager.enumeratorAtURL(cacheURL, includingPropertiesForKeys: resourceKeys, options: .SkipsHiddenFiles, errorHandler: nil)!
            
            var fileURLsToDelete: NSMutableSet = NSMutableSet()
            
            for fileURLString in fileEnumerator {
                
                let fileURL: NSURL = NSURL(string: fileURLString as! String)!
                let resourceValues: NSDictionary = fileURL.resourceValuesForKeys(resourceKeys, error: nil)!
                
                if (resourceValues[NSURLIsDirectoryKey]?.boolValue == true) {
                    continue
                }
                
                // if this image is older than 2 weeks, clear it from cache regardless
                let creationDate: NSDate = resourceValues[NSURLCreationDateKey] as! NSDate
                if  creationDate.laterDate(expirationDate).isEqualToDate(expirationDate) {
                    fileURLsToDelete.addObject(fileURL)
                }
                
                // if this image hasn't been accessed in the last 3 days, remove it from the cache
                let lastAccessDate: NSDate = resourceValues[NSURLContentAccessDateKey] as! NSDate
                let recentlyUsedExpirationDate: NSDate = NSDate().dateByAddingTimeInterval(NSTimeInterval(-self.kMaxCacheAgeForUnusedImages))
                
                if lastAccessDate.laterDate(recentlyUsedExpirationDate).isEqualToDate(recentlyUsedExpirationDate) {
                    if fileURLsToDelete.containsObject(fileURL) == false {
                        fileURLsToDelete.addObject(fileURL)
                    }
                }
                
            }
            
            println("Deleting \(fileURLsToDelete.count) files")
            println("\(fileURLsToDelete)")
            
            for fileURL in fileURLsToDelete {
                self.fileManager.removeItemAtURL(fileURL as! NSURL, error: nil)
            }
            
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                completion()
            })
            
        })
        
    }
    
    private func backgroundCleanDisk() {
        
        println("backgroundCleanDisk")
        /*
        let app: UIApplication = UIApplication.sharedApplication()
        var bgTask: UIBackgroundTaskIdentifier = app.beginBackgroundTaskWithExpirationHandler { () -> Void in
            app.endBackgroundTask(bgTask)
            bgTask = UIBackgroundTaskInvalid
        }
        
        self.cleanDiskWithCompletion { () -> () in
            app.endBackgroundTask(bgTask)
            bgTask = UIBackgroundTaskInvalid
        }
        */
    }
    
}